pkg/logs/logs.go

```diff
@@ -29,103 +35,130 @@ func NewLogsClient(kubernetesClient kclient.ClientInterface) *LogsClient {
 
 var _ Client = (*LogsClient)(nil)
 
-func (o *LogsClient) GetLogsForMode(mode string, componentName string, namespace string, follow bool) ([]map[string]io.ReadCloser, error) {
+func (o *LogsClient) GetLogsForMode(
+	mode string,
+	componentName string,
+	namespace string,
+	follow bool,
+	out io.Writer,
+) (Events, error) {
+	events := Events{
+		Logs: make(chan map[string]interface{}),
+		Err:  make(chan error),
+		Done: make(chan struct{}),
+	}
+
+	go o.getLogsForMode(events, mode, componentName, namespace, follow, out)
+	return events, nil
+}
+
+func (o *LogsClient) getLogsForMode(
+	events Events,
+	mode string,
+	componentName string,
+	namespace string,
+	follow bool,
+	out io.Writer,
+) {
 	var selector string
-	logs := []map[string]io.ReadCloser{}
-	unknownMode := true
+	podChan := make(chan corev1.Pod) // grab the logs of the pod put on this channel
+	errChan := make(chan error)
+	doneChan := make(chan struct{}) // because populating doneChan directly would cause odo logs to exit prematurely.
+
+	go func() {
+		// this go routine gets the logs of the pods put on the podChan
+		for {
+			select {
+			case pod := <-podChan:
+				for _, container := range pod.Spec.Containers {
+					containerLogs, err := o.kubernetesClient.GetPodLogs(pod.Name, container.Name, follow)
+					if err != nil {
+						fmt.Fprintf(out, "failed to get logs for container %s; error: %v", container.Name, err)
```

Comment:
Just curious as to why we don't write this error to `events.Err`..

Reply:
Mainly because I was dicey about stopping the flow if we failed to grab logs of one container among the possibly more than one container that a user could have running. But I have modified it to return an error to `events.Err`. 

We can think if there's any user feedback around what was cooking in my head. 😉

Reply:
Addressed.

Reply:
> Mainly because I was dicey about stopping the flow if we failed to grab logs of one container among the possibly more than one container that a user could have running. But I have modified it to return an error to `events.Err`.
> 
> We can think if there's any user feedback around what was cooking in my head. wink

It made sense to me in the first place because of the other "too many open files" error I mentioned above, and then no new logs were displayed. But okay, let's see if there is any other user feedback on this.